File: app/src/main/java/id/ac/ui/cs/mobileprogramming/nabilafv/rememberyourlord/utils/State.java
package id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.utils;

public class State<T> {
    public State<T> loading() {
        return new Loading<T>();
    }

    public State<T> success(T data) {
        return new Success<T>(data);
    }

    public State<T> failed(Throwable throwable) {
        return new Failed<T>(throwable);
    }

    public State<T> init() {
        return new Initialize<T>();
    }

    public static class Loading<T> extends State{};

    public static class Success<T> extends State{
        public T data;
        public Success(T data){
            this.data = data;
        };
    };

    public static class Failed<T> extends State {
        public Throwable throwable;
        public Failed(Throwable throwable) {
            this.throwable = throwable;
        }
    }

    public static class Initialize<T> extends State{};
}

File: app/src/main/java/id/ac/ui/cs/mobileprogramming/nabilafv/rememberyourlord/utils/State.kt
package id.ac.ui.cs.mobileprogramming.nabilafv.rememberyourlord.utils

open class State<T> {
    class Initialize<T> : State<T>()
    class Loading<T> : State<T>()
    data class Success<T>(val data: T?) : State<T>()
    data class Failed<T>(val throwable: Throwable?) : State<T>()

    fun init(): State<T> {
        return Initialize<T>()
    }

    fun loading(): State<T> {
        return Loading<T>()
    }

    fun success(data: T?): State<T> {
        return Success(data)
    }

    fun failed(throwable: Throwable?): State<T> {
        return Failed<T>(throwable)
    }
}